src/prefect/configuration.py

```diff
@@ -229,7 +229,10 @@ def interpolate_config(config: dict, env_var_prefix: str = None) -> Config:
 
     # interpolate any env vars referenced
     for k, v in list(flat_config.items()):
-        flat_config[k] = interpolate_env_vars(v)
+        val = interpolate_env_vars(v)
+        if isinstance(val, str):
+            val = string_to_type(val)
```

Comment:
This was the key --> if `interpolate_env_vars` returns a string, there's a good chance it's a nested type

Reply:
I was surprised `interpolate_env_vars` wasn't already calling this on the result, and found [this comment](https://github.com/PrefectHQ/prefect/blob/consistent-eval/src/prefect/configuration.py#L85) (which appears to have a bug - we call this on the third pass with `counter > 1`, but the comment says we call it on the second). 

I think now that we're using `literal_eval` this fear is no longer a concern and we could always call `string_to_val(interpolated)` before returning from `interpolate_env_var`, which would guarantee that `interpolate_env_var` always had this expected behavior.

Reply:
Hm ok, should that be called early on (e.g., [right after this line](https://github.com/PrefectHQ/prefect/blob/consistent-eval/src/prefect/configuration.py#L83)) or at the end (e.g., right before [the function returns](https://github.com/PrefectHQ/prefect/blob/consistent-eval/src/prefect/configuration.py#L90)) ?

Reply:
I think right before the function returns, because the interpolation loop will only work on strings. The only real case I can think of where it matters is if you provided a literal list of variables (`'["$A", "$B"]'`), we would want to keep it as an interpretable string for as long as possible. If the first call turned it into `List[str]`, then the substrings would not be expanded further.

Reply:
Discussed IRL; going to clean up the details later and let's go with this implementation!